numpy/lib/index_tricks.py

```diff
@@ -596,6 +597,419 @@ def __iter__(self):
+                    step.append(1 if len(r) <= 1 else (r[-1] - r[0]) // (len(r) - 1))
+                else:
+                    start.append(r)
+                    stop.append(r+1)
+                    step.append(1)
+
+            if singleton:
+                return tuple(start)
+            start = tuple(start)
+            stop = tuple(stop)
+            step = tuple(step)
+        else:
+            # If they are all singletons, then we need to return an index.
+            if all(not isinstance(r, range) for r in new_ranges):
+                return new_ranges
+            # If any of them are ranges, we need to cast all the other
+            # ones to ranges as well
+            start, stop, step = zip(
+                *((r.start, r.stop, r.step) if isinstance(r, range) else (r, r+1, 1)
+                for r in new_ranges))
+
+        return ndrange(start, stop, step)
+
+    def __len__(self):
+        length = 1
+        for r in self._ranges:
+            length = length * len(r)
+        return length
+
+    def __eq__(self, other):
+        # If they don't return the same length tuples, then it is always False
+        if len(self._ranges) != len(other._ranges):
+            return False
+
+        return all(s == o for s, o in zip(self._ranges, other._ranges))
+
+    def __ne__(self, other):
+        # Required for Python 2.7, Python 3 will default to this behaviour
+        # without explicitly defining the __ne__ method.
+        return not self == other
+
+    def __repr__(self):
+        s = "numpy.ndrange({start}, {stop}".format(start=self._start,
+                                                   stop=self._stop)
+        if any(step != 1 for step in self._step):
+            s += ", {step}".format(step=self._step)
+        s += ")"
+        return s
+
+    def __hash__(self):
+        if sys.version_info[0] < 3:
+            raise RuntimeError('numpy.ndrange is hot hashable in Python 2')
+        else:
+            return hash(tuple(hash(r) for r in self._ranges))
+
+    def __array__(self, dtype=None):
+        if dtype is not None:
+            raise ValueError('numpy.ndrange must not be specified.')
+        # How do I assert the dtype is object?
+        arr = empty(self.shape, dtype='object')
+        for i, value in zip(ndrange(self.shape), self):
+            arr[i] = value
+        return arr
+
+    @property
+    def flat(self):
+        # Does flat have to return a flatiter?
+        # Will flatiter even support lazy arrays like this one?
+        return iter(self)
+
+    def ravel(self, order='C'):
+        # Is there a special class I need to return?
+        # Once again, will that special class require the numpy array to be
+        # dense?
+        if order in 'CAK':
+            return iter(self)
+        elif order == 'F':
+            return (i[::-1] for i in itertools.product(*self._ranges[::-1]))
+        else:
+            raise ValueError('order must be one of "C", "F", "A", or "K"')
```

Comment:
Maybe we already discussed this, but why do we want flat and ravel? They seem unnecessary at first glance.

Reply:
@eric-wieser and I were trying to find a way to iterate in `'F'` order that made sense in the context of numpy. Honestly, I don't care about `'F'` order anymore at this point, but this method kinda makes sense.